### ISSUE TITLE
Rust Wasm ABI is unstable and will be changed eventually

### ISSUE THREAD

**Issue by Manishearth** [rust-diplomat/diplomat#661]
Previously: https://github.com/rust-diplomat/diplomat/issues/657 (talking about the specific issue we found)

As explained in https://github.com/rust-lang/rust/issues/129486, the Rust Wasm ABI is currently not following tool conventions and is unstable. https://github.com/rust-lang/rust/pull/117919 details the plan to move forward from here, with a `-Zwasm-c-abi=spec` flag that will become default over time.

This ABI is one where all non-scalar structs are passed as pointers, which means that JS will have to preallocate all structs (except those that wrap a single scalar) and pass them in as pointers. More expensive, but fine.

On the plus side, we can get rid of a lot of the finicky code from https://github.com/rust-diplomat/diplomat/pull/660.

We don't have to do this now, but we should before Rust changes defaults. When we do we should document it. We can potentially support both forms if people really ask for it, but it could get pretty gnarly.

**Comment by sffc** [comment]
Do we have an idea of how `wasm-c-abi=spec` impacts the structs ICU4X actually uses?

Knowing how "crossing the WASM ABI" hasn't seemed to have been an area where enough people really have put enough effort, there might be room to influence the outcome upstream, especially if we can share performance numbers about how much faster certain real-world ICU4X operations might be in the unstable Rust Wasm ABI versus the "spec" Wasm ABI.

**Comment by RalfJung** [comment]
The ABI that rustc currently uses is effectively unsound due to https://github.com/rust-lang/rust/issues/115666. So I don't think we should keep using it under any circumstances. If the wasm ABI can be improved, that should be done as [a wasm-wide change](https://github.com/WebAssembly/tool-conventions/issues/88), not by rustc unilaterally using a different `extern "C"` ABI than the reset of the ecosystem.

**Comment by Manishearth** [comment]
> Do we have an idea of how `wasm-c-abi=spec` impacts the structs ICU4X actually uses?

It basically means that Diplomat slices and ICU4X options bags need to be allocated on the Rust heap before being passed to Rust from JS.

Instead of:

```js
wasm.foo(field1, field2);
```

we will have to do

```js
let buffer = diplomat_alloc(enough space for struct);
ptrWrite(buffer, 0, field1);
ptrWrite(buffer, offset, field2);
wasm.foo(buffer.ptr);
``

There are wasm-wide discussions on doing better here, as Ralf [linked](https://github.com/WebAssembly/tool-conventions/issues/88). We can participate, I'm not sure if we have anything further to bring to the table than what is already being said there.

They seem to be moving towards "pass small structs as fields, pass large structs by pointer", which will kind of be similar to what Rust already does, but with more complexity because of the pointer (whereas currently it's "pass small structs as fields, pass large structs as fields _with padding_").

**Comment by sffc** [comment]
I care slightly less about options bags, but things like DiplomatWriteable and DiplomatSlice are in the critical path for hot functions like format terminals, so I would hope that those don't need an alloc.

I guess a workaround would be to backtrack a bit and inline those structs back to separate parameters in the C functions. Maybe only when the target is WASM.

**Comment by bjorn3** [comment]
You can manually split slices into two arguments when generating the `extern "C"` functions instead of depending on the compiler doing this, right?

**Comment by Manishearth** [comment]
> I guess a workaround would be to backtrack a bit and inline those structs back to separate parameters in the C functions. Maybe only when the target is WASM.

Generally not a great idea since Diplomat's design is that at the C level it's always the same.

And in particular this would mean complicating the macro again (and in a way that's conditional on wasm!), which isn't great.

> You can manually split slices into two arguments when generating the extern "C" functions instead of depending on the compiler doing this, right?

Yes, and we used to do this, but now we consistently use a DiplomatSlice repr(C) struct, which greatly simplifies the code.

So yeah, in theory something we can do, in practice would rather not choose this route.

> but things like DiplomatWriteable and DiplomatSlice are in the critical path for hot functions like format terminals, so I would hope that those don't need an alloc.

Note that allocs aren't _that_ bad here. Allocs are bad in native code, but JS hits the allocator all the time, and while the Rust-side allocator in wasm is not the JS allocator, hitting it is not that expensive compared to other things in JS. I don't think this slows down format functions in a way that's going to be _super_ noticeable. Measurable, probably, noticeable, probably not, it's going to be a same-order-of-magnitude change.

**Comment by Manishearth** [comment]
Either way, the proposed ABI of "small structs are passed directly, large structs indirectly" woudl solve this for us.

**Comment by bjorn3** [comment]
I think you can avoid the allocator by storing everything on the stack instead. You can decrement the stack pointer global before the call, store the arguments at the new stack pointer location, perform the call and restore the stack pointer. This will need the stack pointer to be exported by the wasm module though, or alternatively you need to export a function which adjusts the stack pointer as requested while not touching the stack at all. The latter probably requires hand written wasm.

**Comment by ambiguousname** [comment]
Not sure if anyone has given any more thought to this, but I've been doing some testing with `-Zwasm-c-abi=spec`.

WIP notes are here: https://docs.google.com/document/d/1F0hvY9J8IsRz7wS0allO_E9pu75zKXudDcNWjumx-HA/edit?usp=sharing

For what I've messed with so far, it seems like structs are handled almost the same as we handle them now (padding logic included), with the exception that we are passing structs (and slices!) as allocated pointers, rather than as tuples. There does not seem to be a distinction based on the size of the struct.

Not sure if the ABI is going to change, but I'd be happy to open a PR based on making Diplomat compatible with what exists already. It could be as simple as a tool flag (or detecting the `-Zwasm-c-abi` flag) to switch to new allocation logic.

**Comment by Manishearth** [comment]
I think it should be a tool config (though perhaps we should get our holistic config first: https://github.com/rust-diplomat/diplomat/issues/772).

I would love to see if this can be done with minimal code complexity, otherwise it might be better to just unconditionally move to the new model.

**Comment by ambiguousname** [comment]
I believe minimal complexity is totally doable, so long as no one minds say, a struct having to be allocated in memory every time we want to pass it in as a parameter.

Another solution I can think of is to keep the struct allocated in memory and update it every time a field is changed in JS.

**Comment by Manishearth** [comment]
> a struct having to be allocated in memory every time we want to pass it in as a parameter.

yeah that's inevitable in this model. It's fine. We could optimize this by having some sort of global JS arena so we can reuse buffers in JS. The nice thing is that there's no complicated GC needed: struct allocs are always freed after a call.


> Another solution I can think of is to keep the struct allocated in memory and update it every time a field is changed in JS.

This is the older architecture in the JS backend, I don't like it.

**Comment by bjorn3** [comment]
> For what I've messed with so far, it seems like structs are handled almost the same as we handle them now (padding logic included), with the exception that we are passing structs (and slices!) as allocated pointers, rather than as tuples. There does not seem to be a distinction based on the size of the struct.

Single field structs should be passed the same as their only field: https://github.com/WebAssembly/tool-conventions/blob/main/BasicCABI.md#function-arguments-and-return-values

**Comment by ambiguousname** [comment]
> > For what I've messed with so far, it seems like structs are handled almost the same as we handle them now (padding logic included), with the exception that we are passing structs (and slices!) as allocated pointers, rather than as tuples. There does not seem to be a distinction based on the size of the struct.
> 
> Single field structs should be passed the same as their only field: https://github.com/WebAssembly/tool-conventions/blob/main/BasicCABI.md#function-arguments-and-return-values

Yup! That's the case for the current ABI as well; I forgot to note that down because our tests to catch it were already passing. I'll make sure to update the notes though, thanks!

**Comment by Manishearth** [comment]
IMO go for it wrt implementing configs.

Priority is  a shared config system that can be specified in files, and then we can work on ways to override it.

**Comment by Manishearth** [comment]
Looks like Rust plans to change defaults rather soon (next release)

https://github.com/rust-lang/rust/issues/122532#issuecomment-2723824441


I think the only option available to us is to potentially require nightly flags for wasm, and document how to build wasm stuff on different Rust versions. We can do the internal switchover at any time, though it would be nice to do it after the new stuff has been default and stable for a couple release.


Another solution that was brought up there was that we can start exposing _different_ functions for `target = wasm32-unknown-unknown`: either with the arguments splatted out, or with every struct converted to a reference using indirect passing. Both are tricky since we need to figure out what is and isn't a struct, but it can be done with the right trait magic. Happy to put forward a more concrete design for these if someone is interested in pursuing this route.

If we can make this work as a tool config that would be great, but I'm also okay with the tool having a single route available at any given time. I'm wary of maintaining too much complexity at once.

It would be nice if we could avoid repeated allocation costs by maintaining some global param allocator that contains a preallocated hunk of memory which can grow as needed (though we should cap how much it can grow, since it will not shrink). Indirect params have a very clear lifecycle, so this can basically be a simple bump allocator written in JS, that zeroes after a function call. If we support callbacks we may eventually need something different but that's not a problem for now.

**Comment by ambiguousname** [comment]
I'm in favor of the tool config solution, I think it'd be nice to offer an alternative to whoever's using Diplomat.

I think this approach would be doable, but I'll need to do some more research and testing before I can say that with more confidence.

Since the update is happening soon-ish, I'll make drafting new `conversion.rs` code my top priority.

**Comment by Manishearth** [comment]
Oh, yes, please take your time. And I'm happy to figure out the allocator stuff.

**Comment by workingjubilee** [comment]
@Manishearth You have slightly more time than "next release", it should be next release that it ships to nightly so it will actually be... 3 releases, I think?

**Comment by Manishearth** [comment]
Yep! I understood that part but I probably could have been clearer here!

**Comment by Manishearth** [comment]
@sffc and I discussed this a bit on Slack

Current plan seems to be coalescing around:

 - ICU4X 2.0 (releasing in the next few weeks, hopefully) uses the legacy wasm abi. People should use older stable compilers, or `-Zwasm-c-abi=legacy` on newer nightlies.
 - ICU4X 2.1 will use the new wasm abi. People should use newer stable compilers, or `-Zwasm-c-abi=spec` on older nightlies.
 - We may introduce a `check_abi()` function in diplomat-runtime that returns true when its parameters are as expected, and uses an aggregate parameter, for better diagnostics
 - Diplomat should plan to move over in the next few months, switching to the new ABI around when Rust releases it on stable. We can have the implementation ready beforehand.
 - We _can_ do a config, however I'm worried about buggy behavior being harder to catch when sharing this type of code. Up to @ambiguousname , who was planning on implementing this.

**Comment by Manishearth** [comment]
_Separately_, if we think allocating across FFI is a problem, we can consider implementing some macro solution that produces `foo_wasm` extern C calls that do direct passing.

**Comment by ambiguousname** [comment]
Super happy to try and get a config solution working, my plan is to draft the new bindings system next week. If adding a separate config seems doable from that point, I'd be happy to implement it.

Will link the PR when that gets drafted.
